### PR TITLE
[SPARK-30381][ML] Refactor GBT to reuse treePoints for all trees

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Make GBT reuse splits/treePoints for all trees:
1, reuse splits/treePoints for all trees:
existing impl will find feature splits and transform input vectors to treePoints for each tree; while other famous impls like XGBoost/lightGBM will build a global splits/binned features and reuse them for all trees; 
Note: the sampling rate in existing impl to build `splits` is not the param `subsamplingRate` but the output of `RandomForest.samplesFractionForFindSplits` which depends on `maxBins` and `numExamples`.
Note II: Existing impl do not guarantee that splits among iteration are the same, so this may cause a little difference in convergence.

2, do not cache input vectors:
existing impl will cached the input twice: 1,`input: RDD[Instance]` is used to compute/update prediction and errors; 2, at each iteration, input is transformed to bagged points, the bagged points will be cached during this iteration;
In this PR,`input: RDD[Instance]` is no longer cached, since it is only used three times: 1, compute metadata; 2, find splits; 3, converted to treePoints;
Instead, the treePoints `RDD[TreePoint]` is cached, at each iter, it is convert to bagged points by attaching extra `labelWithCounts: RDD[(Double, Int)]` containing residuals/sampleCount information, this rdd is relative small (like cached `norms` in KMeans);
To compute/update prediction and errors, new prediction method based on binned features are added in `Node`

### Why are the changes needed?
for perfermance improvement: 
1,40%~50% faster than existing impl
2,save 30%~50% RAM

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
existing testsuites & several manual tests in REPL

### CODE REVIEW

**Comment by zhengruifeng** [comment]
testcode:
```scala
import org.apache.spark.ml.regression._
import org.apache.spark.storage.StorageLevel

var df = spark.read.format("libsvm").load("/data1/Datasets/a9a/a9a")

(0 until 8).foreach{ _ => df = df.union(df) }
df.persist(StorageLevel.MEMORY_AND_DISK)

df.count
df.count
df.count

val gbt = new GBTRegressor().setMaxIter(10)
val gbtm = gbt.fit(df)


val start = System.currentTimeMillis; val gbtm = gbt.fit(df); val end = System.currentTimeMillis; end - start

gbtm.evaluateEachIteration(df, "squared")


val gbt2 = new GBTRegressor().setMaxIter(10).setSubsamplingRate(0.8)

val start = System.currentTimeMillis; val gbtm2 = gbt2.fit(df); val end = System.currentTimeMillis; end - start

gbtm2.evaluateEachIteration(df, "squared")
```

result:
about 48% faster than existing impl
|dur_gbt(new) | dur_gbt2(new) | dur_gbt(old) | dur_gbt2(old) |
|------|----------|------------|----------|
|197777|188205|133214|134787|

|loss(new) | loss2(new) | loss(old) | loss2(old) |
|------|----------|------------|----------|
|0.4283679582338368|0.42678864636469305|0.4283679582338368|0.42678864636469305|

The convergences of `gbt` should be the same, since `splits` are always built on the whole dataset and should be the same.
The convergences of `gbt2` do not have to be the same. However in above tests they happen to be the same, maybe due to that the input df is generated by repeate `a9a` 1024 times.

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala [hunk 1]
```diff
@@ -111,14 +113,18 @@ private[spark] object GradientBoostedTrees extends Logging {
...
       initTree: DecisionTreeRegressionModel,
-      loss: OldLoss): RDD[(Double, Double)] = {
-    data.map { case Instance(label, _, features) =>
-      val pred = updatePrediction(features, 0.0, initTree, initTreeWeight)
-      val error = loss.computeError(pred, label)
-      (pred, error)
+      loss: OldLoss,
+      bcSplits: Broadcast[Array[Array[Split]]]): RDD[(Double, Double)] = {
+    data.mapPartitions { iter =>
+      val splits = bcSplits.value
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: You don't really need mapPartitions for this. `.value` just returns a local field always

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala [hunk 2]
```diff
@@ -313,29 +335,70 @@ private[spark] object GradientBoostedTrees extends Logging {
...
     val firstTreeWeight = 1.0
     baseLearners(0) = firstTreeModel
     baseLearnerWeights(0) = firstTreeWeight
 
-    var predError = computeInitialPredictionAndError(input, firstTreeWeight, firstTreeModel, loss)
+    var predError = computeInitialPredictionAndError(
+      treePoints, firstTreeWeight, firstTreeModel, loss, bcSplits)
     predErrorCheckpointer.update(predError)
-    logDebug("error of gbt = " + computeWeightedError(input, predError))
+    logDebug("error of gbt = " + computeWeightedError(treePoints, predError))
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: Nit, while you're here, add interpolation on debug logs that you change?

**Review by srowen — COMMENTED** [review]
It's a little hard to evaluate, but I don't see any issues with it on reading the code. If the tests pass and it's faster, I can see the advantage here.

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala (hunk 2 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: Sorry I do not get the point, do you mean checking the "error of gbt" in the log?

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: I had checked the convergence by `gbtm.evaluateEachIteration(df, "xxx")` and the curves are similar (may differ due to different `splits`)

Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: No I just mean use `s"..."` string interpolation

**Review by srowen — APPROVED** [review]

**Comment by zhengruifeng** [comment]
Merged to master, thanks @srowen for reviewing!
